src/client.rs

```diff
@@ -136,36 +96,195 @@ impl Client {
+    }
+}
+
+type Promise =
+    std::pin::Pin<Box<dyn std::future::Future<Output = Result<Arc<Certificates>, Error>>>>;
+
+enum RefreshState {
+    Ready(Arc<Certificates>),
+    Expired(Shared<Promise>),
+    Uninitialized,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+struct CertsObject {
+    keys: Vec<Cert>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+struct Cert {
+    kid: String,
+    e: String,
+    kty: String,
+    alg: String,
+    n: String,
+    r#use: String,
+}
+
+type Key = String;
+
+#[derive(Clone)]
+pub struct Certificates {
+    keys: BTreeMap<Key, Cert>,
+    pub expiry: Option<Instant>,
+}
+
+impl Certificates {
+    pub fn new() -> Self {
+        Self {
+            keys: BTreeMap::new(),
+            expiry: None,
         }
+    }
+
+    /// Downloads the public Google certificates even if the current certificates have not expired.
+    pub async fn get(client: &Client) -> Result<Certificates, Error> {
+        Certificates::get_with_http_client(&client.client).await
+    }
 
+    async fn get_with_http_client(client: &HttpClient) -> Result<Certificates, Error> {
+        const URL: &str = "https://www.googleapis.com/oauth2/v2/certs";
+
+        let url = URL.parse().unwrap();
+        let response = client.get(url).await?;
+        let expiry = response
+            .headers()
+            .get("Cache-Control")
+            .and_then(|val| val.to_str().ok())
+            .and_then(cache_control::CacheControl::from_value)
+            .and_then(|cc| cc.max_age)
+            .and_then(|max_age| {
+                let seconds = max_age.num_seconds();
+                if seconds >= 0 {
+                    Some(Instant::now() + Duration::from_secs(seconds as u64))
+                } else {
+                    None
+                }
+            });
         let body = hyper::body::aggregate(response).await?;
-        Ok(serde_json::from_reader(body.reader())?)
+        let certs: CertsObject = serde_json::from_reader(body.reader())?;
+        let mut keys = BTreeMap::new();
+        for cert in certs.keys {
+            keys.insert(cert.kid.clone(), cert);
+        }
+        Ok(Certificates { keys, expiry })
+    }
+
+    /// Downloads the public Google certificates if it didn't do so already, or based on expiry of
+    /// their Cache-Control. Returns `true` if the certificates were updated.
+    pub async fn refresh(&mut self) -> Result<bool, Error> {
+        if !self.is_expired() {
+            return Ok(false);
+        }
+
+        let client = Client::new();
+        *self = Certificates::get(&client).await?;
+        Ok(true)
+    }
+
+    /// Returns true if all cached certificates are expired (or if there are no cached certificates).
+    pub fn is_expired(&self) -> bool {
+        match self.expiry {
+            Some(expiry) => expiry <= Instant::now(),
+            None => true,
+        }
+    }
+
+    fn get_range<'a>(&'a self, kid: &Option<String>) -> Result<Range<'a, Key, Cert>, Error> {
```

Comment:
It seems to me the range part of this functionality isn't actually used, but I might be missing future proofing? Else the BTreeMap could be swapped with a HashMap.